src/main/java/org/opensearch/security/auth/BackendRegistry.java

```diff
@@ -415,6 +404,22 @@ public boolean authenticate(final SecurityRequestChannel request) {
                 }
             }
 
+            if (authCredentials == null
+                && anonymousAuthEnabled
+                && isRequestForAnonymousLogin(request.params())
+                && !checkIfRequestContainsBasicAuthHeader(request.getHeaders())) {
```

Comment:
I believe some authentication providers can provider information by the query parameter.  IMO I would rather check all authentication provides first, then if none have found the user, then only if `anonymousAuthEnabled` swallow the 401s.

Alternatively, if there was an opt-in header `X-Authorization: Anonymous`, then that would be a viable signal IMO.  What do you think?

Reply:
> I believe some authentication providers can provider information by the query parameter. IMO I would rather check all authentication provides first, then if none have found the user, then only if anonymousAuthEnabled swallow the 401s.

This is how it behaves today. It loops over all auth domains, and if authCredentials are still null, even after challenging authenticator check, then it assumes anonymous role if no authorization header was present in the first place.

> Alternatively, if there was an opt-in header `X-Authorization: Anonymous`, then that would be a viable signal IMO. What do you think?

I've tried with something similar but the requests failed when a call was made to get dashboards configuration via `GET /.kibana/config`. The reason was that the authorization header could not be passed from `plugin/security/authinfo` to the subsequent call to get kibana config from the dashboard as this happens in a one flow after clicking `Log in as Anonymous`.  https://github.com/opensearch-project/security/pull/4097#issuecomment-2004657061.

IMO this approach addresses both bugs with minimalistic change. Are there any concerns with this approach?

Reply:
The check `headers.containsKey(HttpHeaders.AUTHORIZATION)` is not authoritative to indicate that the caller isn't attempting to authenticate.  Query parameters can be used or alternative headers - this is not a reliable with existing auth providers and creates a sharp edge where there could be unexpected behavior with future authentication providers such as SigV4 for AWS' ecosystem or others.

I'm recommend working backwards from if there is a preferred authentication mechanism how can anonymous authentication also be allowed, where existing behavior around receiving 401s needs to be preserved.

IMO there is _no_ way to do this correctly without the client performing explicit negotiation to include a header.

Reply:
This case only concerns Basic Auth. I'm trying to resolve: https://github.com/opensearch-project/security-dashboards-plugin/issues/1840 where upon passing wrong credentials user was automatically logged in as anonymous. For other types of authorization, like SigV4, if authn fails it won't redirect to opensearch, if authn succeeds then the `authCredentials == null` check will fail and anonymous login check will not be reached.

Reply:
@peternied I am trying to understand your perspective here, what is the difference between checking for a `auth_request_type` header, and a `X-Authorization` header? I do agree that the logic should be that anonymous auth should be checked last if possible.

Reply:
I just re-read the code - `isRequestForAnonymousLogin(...)` does I was looking sorry for the confusion.